app/src/main/java/com/nervousfish/nervousfish/activities/VisualVerificationActivity.java

```diff
@@ -59,7 +59,7 @@ private void nextActivity() {
             LOGGER.info("Sending my profile with name: " + myProfile.getName() + ", public key: "
                     + myProfile.getPublicKey().toString());
             final Contact myProfileAsContact = new Contact(myProfile.getName(), new SimpleKey("simplekey", "73890ien"));
-            this.serviceLocator.getBluetoothHandler().send(myProfileAsContact);
+            this.serviceLocator.getBluetoothHandler().send(this.serviceLocator.getBluetoothHandler().objectToBytes(myProfileAsContact));
```

Comment:
> Not sure if you are aware of this, but you're actually introducing duplicate code because now you constantly have to convert these things to bytes separately. I don't know where the other duplicate code would appear since the problem you're describing is based in a single class. Can you explain where we would have had duplicate code if it where not for this change?
Yes, I can. NFC needs the message as bytes in order to send it. However, it cannot use the send() method of APairingHandler, because it has to send the bytes in the activity itself. Because with the current implementation it is impossible to convert an object to bytes without duplicating the code in the send() function of APairingHandler, we want to expose that functionality in APairingHandler to classes like NFC.

> Indeed a single task, send data. I don't care what data, is it a string? send it. Is it an object? send it. I don't care how it has to be send... Ow it has to be send as a byte-array? Well let the send() method handle that.

You could look at it that way, but **from the view of NFCActivity** the method does convert the object to bytes (great, love that, I need that functinality!) AND the method sends the bytes using the send method implemented by Bluetooth, NFC or QRHandler (Uh oh, I don't want that! It's Android so NFCHandler cannot do that at all, I have to do that in the NFC activity myself)

> OK, sure. So we need a send method that accepts byte-arrays and one that accepts object? That works for me...

Sadly, I thought about that as well, but it is not possible it seems. The send method doesn't return anything, but the method createNdefMessage (which MUST be implemented and is called by Android) returns and NdefMessage and send doesn't return such a message of course. Bluetooth doesn't need that, QR doesn't need an NdefMessage, only NFC. Therefore we cannot delegate the functionality of createNdefMessage to NFCHandler and therefore we cannot use NFCHandler at all.